File: java/Lists/CircularLinkedList.java
package Lists;

import definitions.LinkedLists;

import java.util.Iterator;

public class CircularLinkedList<T> implements Iterable<T>, LinkedLists<T> {

    private SingleLinkedList.SingleLinkedListNode<T> head, tail;
    private int size = 0;

    private void init(T val) {
        var node = new SingleLinkedList.SingleLinkedListNode<T>(val);
        head = node;
        tail = head;
        size = 1;
    }

    @Override
    public boolean isEmpty() {
        return size == 0 && head == null && tail == null;
    }

    @Override
    public void add(T val) {
        addLast(val);
    }

    @Override
    public void add(int index, T val) {
        if (!isValidInsertIndex(index)) return;
        if (index == size) {
            addLast(val);
            return;
        }
        var temp = head;
        for (int i = 0; i < index - 1; i++) {
            temp = temp.next;
        }
        var next = temp.next;
        var node = new SingleLinkedList.SingleLinkedListNode<>(val, next);
        temp.next = node;
        size++;
    }

    @Override
    public void addFirst(T val) {
        if (isEmpty()) {
            init(val);
            return;
        }
        var node = new SingleLinkedList.SingleLinkedListNode<T>(val, head);
        tail.next = node;
        node.next = head;
        head = node;
        size++;
    }

    @Override
    public void addLast(T val) {
        if (isEmpty()) {
            init(val);
            return;
        }
        var node = new SingleLinkedList.SingleLinkedListNode(val, head);
        tail.next = node;
        node.next = head;
        tail = node;
        size++;
    }

    @Override
    public void clear() {
        head = null;
        tail = null;
        size = 0;
    }

    @Override
    public T element() {
        return getFirst();
    }

    private boolean isValidInsertIndex(int index) {
        return index > 0 && index <= size;
    }

    private boolean isValidGetIndex(int index) {
        return index > 0 && index < size;
    }

    @Override
    public T get(int index) {
        if (isEmpty() || !isValidGetIndex(index)) return null;
        if (size - 1 == index) return getLast();
        var temp = head;
        for (int i = 0; i < index; i++) {
            temp = temp.next;
        }
        return temp.val;
    }

    @Override
    public T getFirst() {
        if (isEmpty()) return null;
        return head.val;
    }

    @Override
    public T getLast() {
        if (isEmpty()) return null;
        return tail.val;
    }

    @Override
    public int indexOf(T val) {
        var temp = head;
        int i = 0;
        do {
            if (temp.val.equals(val)) return i;
            temp = temp.next;
            i++;
        } while (temp != head);
        return -1;
    }

    @Override
    public int lastIndexOf(T val) {
        var temp = head;
        int i = 0;
        int foundIndex = -1;
        do {
            if (temp.val.equals(val)) foundIndex = i;
            temp = temp.next;
            i++;
        } while (temp != null);
        return foundIndex;
    }

    @Override
    public T remove() {
        return removeFirst();
    }

    @Override
    public T remove(int index) {
        if (!isValidGetIndex(index)) return null;
        var node = head;
        while (index - 1 > 0) {
            node = node.next;
            index--;
        }
        SingleLinkedList.SingleLinkedListNode next = null;
        if (node.next != null) next = node.next.next;
        node.next = next;
        size--;
        return null;
    }

    @Override
    public T removeFirst() {
        if (isEmpty()) return null;
        var val = head.val;
        head = head.next;
        if (head == null) {
            clear();
            return val;
        }
        tail = head;
        size--;
        return val;
    }

    @Override
    public T removeLast() {
        if (isEmpty()) return null;
        var val = tail.val;
        if (head == tail) {
            clear();
            return val;
        }
        var temp = head;
        while (temp.next != tail) {
            temp = temp.next;
        }
        temp.next = head;
        size--;
        return val;
    }

    @Override
    public void set(int index, T val) {
        if (!isValidGetIndex(index)) return;
        var node = head;
        while (index > 0) {
            node = node.next;
            index--;
        }
        node.val = val;

    }

    @Override
    public int size() {
        return this.size;
    }

    public class CircularLinkedListItr<T> implements Iterator<T> {
        private SingleLinkedList.SingleLinkedListNode node = CircularLinkedList.this.head;
        int attempt = 0;

        @Override
        public boolean hasNext() {
            return attempt < CircularLinkedList.this.size;
        }

        @Override
        public T next() {
            var val = (T) node.val;
            node = node.next;
            attempt++;
            return val;
        }
    }

    @Override
    public Iterator<T> iterator() {
        return null;
    }
}

File: typescript/src/lists/circular-linked-list.ts
import { SingleLinkNode } from "./single-linked-list.ts";

export default class SingleLinkedList<T> {
  private head: SingleLinkNode<T> | null;
  private tail: SingleLinkNode<T> | null;
  private count: number;

  public constructor() {
    this.head = null;
    this.tail = null;
    this.count = 0;
  }

  public get size() {
    return this.count;
  }

  private isValidGetIndex(index: number) {
    return index >= 0 && index < this.size;
  }

  private isValidAddIndex(index: number) {
    return index >= 0 && index <= this.size;
  }

  private init(val: T) {
    this.head = new SingleLinkNode(val);
    this.tail = this.head;
    this.tail!.next = this.head;
    this.count = 1;
  }

  public addFirst(val: T) {
    if (this.size == 0) return this.init(val);
    const node = new SingleLinkNode(val);
    node.next = this.head;
    this.tail!.next = node;
    this.head = node;
    this.count++;
  }

  public addLast(val: T) {
    if (this.size == 0) return this.init(val);
    const node = new SingleLinkNode(val);
    this.tail!.next = node;
    node.next = this.head;
    this.tail = node;
    this.count++;
  }

  public add(val: T, index?: number) {
    index ??= this.size;
    if (!this.isValidAddIndex(index)) return;
    if (index == 0) return this.addFirst(val);
    if (index == this.size) return this.addLast(val);
    let temp = this.head;
    while (index - 1 > 0) {
      temp = temp!.next;
      index--;
    }
    const node = new SingleLinkNode(val);
    node.next = temp!.next;
    temp!.next = node;
    this.count++;
  }

  public getFirst() {
    if (this.size == 0) return null;
    return this.head!.val;
  }

  public getLast() {
    if (this.size == 0) return null;
    return this.tail!.val;
  }

  public get(index: number) {
    if (!this.isValidGetIndex(index)) return null;
    if (index == 0) return this.getFirst();
    if (index == this.size - 1) return this.getLast();
    let temp = this.head;
    while (index > 0) {
      temp = temp!.next;
      index--;
    }
    return temp!.val;
  }

  public removeFirst() {
    if (this.size == 0) return null;
    const val = this.head!.val;
    this.head != this.head!.next;
    if (this.head == null) {
      this.clear();
      return val;
    }
    this.tail!.next = this.head;
    this.count--;
    return val;
  }

  public removeLast() {
    if (this.size == 0) return null;
    const val = this.tail!.val;
    if (this.size == 1) {
      this.clear();
      return val;
    }
    let temp = this.head;
    while (temp!.next != this.tail) {
      temp = temp!.next;
    }
    temp!.next = this.head;
    this.count--;
    return val;
  }

  public remove(index?: number) {
    index ??= this.size - 1;
    if (!this.isValidGetIndex(index)) return null;
    if (index == 0) return this.removeFirst();
    if (index == this.size - 1) return this.removeLast();
    let temp = this.head;
    while (index > 0) {
      temp = temp!.next;
      index--;
    }
    temp!.next = temp!.next!.next;
    this.count--;
  }

  public set(index: number, val: T) {
    if (!this.isValidGetIndex(index)) return;
    let temp = this.head;
    while (index > 0) {
      temp = temp!.next;
      index--;
    }
    temp!.val = val;
  }

  public includes(val: T) {
    let temp = this.head;
    while (temp != null) {
      if (val == temp!.val) return true;
      temp = temp!.next;
    }
    return false;
  }

  public clear() {
    this.head = null;
    this.tail = null;
    this.count = 0;
  }

  public toArray() {
    const arr: T[] = [];
    if (this.size == 0) return arr;
    arr.push(this.head!.val);
    if (this.size == 1) {
      return arr;
    }
    let temp = this.head!.next;
    while (temp != this.head) {
      arr.push(temp!.val);
      temp = temp!.next;
    }
    return arr;
  }
}
